File: python/detect.py
#!/usr/bin/env python3
"""File Encoding Detector — Python implementation.

Detects the character encoding of every file in a given folder.

Usage:
    python detect.py <folder> [options]

See README.md or run with --help for full usage.
"""

from __future__ import annotations

import argparse
import sys
from datetime import datetime
from pathlib import Path

from lib.detector import detect_file
from lib.exporter import export_csv, export_json, print_console
from lib.scanner import iter_files


# ---------------------------------------------------------------------------
# CLI argument parsing
# ---------------------------------------------------------------------------

def build_parser() -> argparse.ArgumentParser:
    """Build and return the argument parser."""
    parser = argparse.ArgumentParser(
        prog="detect",
        description="Detect the character encoding of files in a folder.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
examples:
  python detect.py C:\\Users\\me\\documents
  python detect.py ./src --recursive
  python detect.py ./data --ext .csv .txt --output csv
  python detect.py ./project --recursive --output json --file results.json
""",
    )

    parser.add_argument(
        "folder",
        type=Path,
        help="Path to the folder to scan.",
    )
    parser.add_argument(
        "--recursive", "-r",
        action="store_true",
        default=False,
        help="Scan subdirectories recursively (default: off).",
    )
    parser.add_argument(
        "--output", "-o",
        choices=["console", "csv", "json"],
        default="console",
        help="Output format (default: console).",
    )
    parser.add_argument(
        "--file", "-f",
        type=Path,
        default=None,
        help="Destination file for CSV/JSON export. Auto-named if not specified.",
    )
    parser.add_argument(
        "--hidden",
        action="store_true",
        default=False,
        help="Include hidden files and system folders (default: off).",
    )
    parser.add_argument(
        "--ext",
        nargs="+",
        metavar="EXT",
        default=None,
        help="Only scan files with these extension(s), e.g. --ext .txt .csv",
    )
    parser.add_argument(
        "--min-size",
        type=int,
        default=0,
        metavar="BYTES",
        help="Skip files smaller than N bytes (default: 0).",
    )
    parser.add_argument(
        "--max-size",
        type=int,
        default=None,
        metavar="BYTES",
        help="Skip files larger than N bytes (default: no limit).",
    )

    return parser


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _auto_output_path(folder: Path, fmt: str) -> Path:
    """Generate an auto-named output file path in the current directory."""
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    stem = folder.resolve().name or "scan"
    return Path(f"{stem}_encoding_{timestamp}.{fmt}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> int:
    """Entry point. Returns an exit code."""
    parser = build_parser()
    args = parser.parse_args()

    # Validate folder
    folder: Path = args.folder.expanduser().resolve()
    if not folder.exists():
        print(f"error: folder not found: {folder}", file=sys.stderr)
        return 2
    if not folder.is_dir():
        print(f"error: path is not a directory: {folder}", file=sys.stderr)
        return 2

    # Collect files
    files = list(
        iter_files(
            folder,
            recursive=args.recursive,
            include_hidden=args.hidden,
            extensions=args.ext,
            min_size=args.min_size,
            max_size=args.max_size,
        )
    )

    if not files:
        print("No files found matching the given criteria.", file=sys.stderr)
        return 0

    print(f"Scanning {len(files)} file(s) in {folder} …", file=sys.stderr)

    # Detect encodings
    results = [detect_file(f) for f in files]

    # Output
    if args.output == "console":
        print_console(results, root=folder)

    elif args.output == "csv":
        out_path = args.file or _auto_output_path(folder, "csv")
        try:
            export_csv(results, out_path, root=folder)
        except OSError as e:
            print(f"error: could not write CSV: {e}", file=sys.stderr)
            return 3
        # Also print a summary to console
        print_console(results, root=folder)

    elif args.output == "json":
        out_path = args.file or _auto_output_path(folder, "json")
        try:
            export_json(results, out_path, root=folder)
        except OSError as e:
            print(f"error: could not write JSON: {e}", file=sys.stderr)
            return 3
        print_console(results, root=folder)

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: nodejs/bin/detect.js
#!/usr/bin/env node
/**
 * @file detect.js
 * CLI entry point for the File Encoding Detector (Node.js).
 *
 * Usage:
 *   node bin/detect.js <folder> [options]
 *   npx file-encoding-detector <folder> [options]
 */

import chalk from "chalk";
import { program } from "commander";
import { existsSync, statSync } from "fs";
import { resolve } from "path";

import { detectFile } from "../lib/detector.js";
import { exportCsv, exportJson, printConsole } from "../lib/exporter.js";
import { collectFiles } from "../lib/scanner.js";

// ---------------------------------------------------------------------------
// CLI definition
// ---------------------------------------------------------------------------

program
  .name("detect-encoding")
  .description("Detect the character encoding of every file in a folder.")
  .version("1.0.0")
  .argument("<folder>", "Path to the folder to scan")
  .option("-r, --recursive", "Scan subdirectories recursively", false)
  .option(
    "-o, --output <format>",
    "Output format: console, csv, json",
    "console",
  )
  .option(
    "-f, --file <path>",
    "Destination file for CSV/JSON export (auto-named if omitted)",
  )
  .option("--hidden", "Include hidden files and system folders", false)
  .option(
    "--ext <exts...>",
    "Only scan files with given extension(s), e.g. --ext .txt .csv",
  )
  .option(
    "--min-size <bytes>",
    "Skip files smaller than N bytes",
    (v) => parseInt(v, 10),
    0,
  )
  .option(
    "--max-size <bytes>",
    "Skip files larger than N bytes",
    (v) => parseInt(v, 10),
    null,
  )
  .option(
    "-c, --concurrency <n>",
    "Number of files to process in parallel",
    (v) => parseInt(v, 10),
    10,
  )
  .addHelpText(
    "after",
    `
Examples:
  node detect.js C:\\Users\\me\\documents
  node detect.js ./src --recursive
  node detect.js ./data --ext .csv .txt --output csv
  node detect.js ./project --recursive --output json --file results.json
`,
  );

program.parse();

const opts = program.opts();
const folder = program.args[0];

// ---------------------------------------------------------------------------
// Validate folder
// ---------------------------------------------------------------------------

const absFolder = resolve(folder);

if (!existsSync(absFolder)) {
  console.error(chalk.red(`error: folder not found: ${absFolder}`));
  process.exit(2);
}
if (!statSync(absFolder).isDirectory()) {
  console.error(chalk.red(`error: path is not a directory: ${absFolder}`));
  process.exit(2);
}

// ---------------------------------------------------------------------------
// Validate output format
// ---------------------------------------------------------------------------

const validFormats = ["console", "csv", "json"];
if (!validFormats.includes(opts.output)) {
  console.error(
    chalk.red(
      `error: invalid output format "${opts.output}". Choose: console, csv, json`,
    ),
  );
  process.exit(1);
}

// ---------------------------------------------------------------------------
// Auto-name output file
// ---------------------------------------------------------------------------

function autoOutputPath(fmt) {
  const now = new Date();
  const ts =
    now.getFullYear().toString() +
    String(now.getMonth() + 1).padStart(2, "0") +
    String(now.getDate()).padStart(2, "0") +
    "_" +
    String(now.getHours()).padStart(2, "0") +
    String(now.getMinutes()).padStart(2, "0") +
    String(now.getSeconds()).padStart(2, "0");
  const stem = absFolder.split(/[\\/]/).pop() || "scan";
  return `${stem}_encoding_${ts}.${fmt}`;
}

// ---------------------------------------------------------------------------
// Concurrent processing with a semaphore
// ---------------------------------------------------------------------------

/**
 * Process an array of tasks with a capped concurrency.
 *
 * @template T
 * @param {T[]}              items
 * @param {(item: T) => Promise<unknown>} fn
 * @param {number}           limit
 */
async function withConcurrency(items, fn, limit) {
  const results = [];
  let index = 0;

  async function worker() {
    while (index < items.length) {
      const i = index++;
      results[i] = await fn(items[i]);
    }
  }

  const workers = Array.from({ length: Math.min(limit, items.length) }, worker);
  await Promise.all(workers);
  return results;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

(async () => {
  // Collect files
  const files = await collectFiles(absFolder, {
    recursive: opts.recursive,
    includeHidden: opts.hidden,
    extensions: opts.ext ?? null,
    minSize: opts.minSize,
    maxSize: opts.maxSize,
  });

  if (files.length === 0) {
    console.error(chalk.yellow("No files found matching the given criteria."));
    process.exit(0);
  }

  process.stderr.write(
    chalk.dim(`Scanning ${files.length} file(s) in ${absFolder} …\n`),
  );

  // Detect encodings (with concurrency limit)
  const results = await withConcurrency(files, detectFile, opts.concurrency);

  // Output
  if (opts.output === "console") {
    printConsole(results, absFolder, chalk);
  }

  if (opts.output === "csv") {
    const outPath = opts.file ?? autoOutputPath("csv");
    try {
      await exportCsv(results, outPath, absFolder);
    } catch (err) {
      console.error(chalk.red(`error: could not write CSV: ${err.message}`));
      process.exit(3);
    }
    printConsole(results, absFolder, chalk);
  }

  if (opts.output === "json") {
    const outPath = opts.file ?? autoOutputPath("json");
    try {
      await exportJson(results, outPath, absFolder);
    } catch (err) {
      console.error(chalk.red(`error: could not write JSON: ${err.message}`));
      process.exit(3);
    }
    printConsole(results, absFolder, chalk);
  }

  process.exit(0);
})();
